perf(pathPlanner): path planner performance optimization

AhmadAlamoush opened the pull request:
Reduce the size of the learning units and skills to enhance performance
- Just consider in-scope learning units (Starting from the goal backward)
- Just consider in-scope skills (Just the skills involved with the in-scope learning units)

AhmadAlamoush committed:
perf(pathPlanner): path planner performance optimization

Reduce the size of the learning units and skills to enhance performance
- Just consider in-scope learning units (Starting from the goal backward)
- Just consider in-scope skills (Just the skills involved with the in-scope learning units)

AhmadAlamoush committed:
perf(pathPlanner): path planner performance optimization

Fix the format of the files.

Elscha commented on src/lib/pathPlanner.spec.ts:
```diff
@@ -1544,6 +1564,108 @@ describe("Path Planer", () => {
 		});
 	});
+
+	describe("filter out of scope learning units", () => {
```
Please add (in a separate describe block) some performance tests of cases were we know that the old approach could no handle this in adequate time.

For this measure the time before the planing and directly after the planning and check in an expect that the number is smaller than a expected time. Consider for this that the build server may has a worese hardware, that future changes require additional time, etc.. Thus, add some additional time for your test, but in total it should be still less time than the current approach (e.g., 2 sec instead of 25 sec than use less than 5 sec).

An example is provided in [nm-skill-service/src/learningPath/learningPath.service.spec.ts L:735 - L:759](https://github.com/e-Learning-by-SSE/nm-skill-service/blob/main/src/learningPath/learningPath.service.spec.ts#L735-L759)

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -149,7 +152,7 @@ export function getPaths<LU extends LearningUnit>({
 	// Filter LearningUnits which cannot reach the goal
 	const goalIds = goal.map(skill => skill.id);
-	const filteredLearningUnits = learningUnits.filter(
+	const filteredLearningUnits = inScopeLearningUnits.filter(
```
Are you aware wehter we still need this. This basically filters LearningUnits that do not leat to any part of the goal. this should probably be supersseded by your changes.

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+				learningUnit.teachingGoals.some(skill => skill.id == nestedSkill)
+			);
+			const luNotExist = lus.filter(lu => !filteredLearningUnits.includes(lu));
+			filteredLearningUnits.push(...luNotExist);
```
Looks very complicated and with low performance for me. Would be a Set a better option. This already checks that elements are not copied twice.

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+		lu.teachingGoals.some(skill => goal.includes(skill))
+	);
+
+	goal.forEach(skill => {
```
1.) Please explain the approach your idea for this block with a small comment.

2.) This function seems to be incomplete, as:
    - Only nested skills, but not parent/skills without children are checked. Are they also covered?
    - Not recursive. What happens with children of children

For 2.) Please consider the following situations.
1.) Goal: `[Java]` with (parent -> children): `Java `-> `Control Structures` -> `Branching `-> `If-Concept` -> `[if, else, nested if, complex conditions, ...]`
2.) Goal: `[if, else] `

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
 	return skillAnalyzedPath;
 }
+
+export function filterOutOfScopeLus<LU extends LearningUnit>(
```
JSDoc including a brief idea of the algorithmic

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+		) {
+			continue;
+		}
+		let requiredLearningUnit = learningUnits.filter(lu =>
```
It seems that you nor re-assign the variable (but push additional elements into the array). Than const should till be available.

Please add a small comment for what this variable is used for.

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+			lu.teachingGoals.some(skill => learningUnit.requiredSkills.includes(skill))
+		);
+
+		learningUnit.requiredSkills.forEach(skill => {
```
Again no recursion of nested skills, why no top skills or nested of nested skills? Please check if this is realy works for all cases

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+		);
+
+		learningUnit.requiredSkills.forEach(skill => {
+			requiredLearningUnit.push(
```
Is it possible to make it more readable by avoding this side-effect or by spliting it into 2 statements?

For instance:
```typescript
const predecessorLUs = learningUnit.requiredSkills.forEach(skill => {
    learningUnits.filter(...)
                 .filter(...)
});
requiredLearningUnit.push(predecessorLUs);
```

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+			);
+		});
+
+		const luNotExist = requiredLearningUnit.filter(lu => !inScopeLearningUnits.includes(lu));
+
+		filteredLearningUnits.push(...luNotExist);
+		inScopeLearningUnits.push(...luNotExist);
```
What happens here?

Please write a comment

Elscha commented on src/lib/pathPlanner.ts:
```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
+	return inScopeLearningUnits;
+}
+
+export function filterOutOfScopeSkills<LU extends LearningUnit>(
```
Please add JSDoc

Is here recursion for nested skills needed?

Elscha commented:
Please add a comparison / benchmark of the previous and the new approach.

AhmadAlamoush committed:
Merge branch 'main' of https://github.com/e-Learning-by-SSE/nm-skill-lib into fix-learning-path-optimization

AhmadAlamoush commented:
> new approach

<img width="693" alt="ApproachesComparison" src="https://github.com/e-Learning-by-SSE/nm-skill-lib/assets/75734718/bd9bc4cc-8331-40f5-bd00-387bfd6872ff">

AhmadAlamoush committed:
perf(pathPlanner): path planner performance optimization

Change the way to filter the learning units. Looping over the required skills instead of looping over learning units.
Taking in consideration the skill parent/children relationship.

Elscha committed:
perf(pathPlanner): Directly use of a set while filtering for useful LUs

Benefits:
* Avoids unnecessary creation of collection: array -> set -> array
* Filtering for duplicates LUs early

Elscha committed:
perf(pathPlanner): Directly use of a set while filtering for useful LUs

Benefits:
* Avoids unnecessary creation of collection: array -> set -> array
* Filtering for duplicates LUs early

AhmadAlamoush committed:
perf(pathPlanner): Apply directly use of a set for skill filtering

* apply the use of a set in function filterOutOfScopeSkills

AhmadAlamoush committed:
perf(pathPlanner): Use processed skill set to avoid reprocessing the skills
Benefits:
* Avoids unnecessary reprocessing of already processed skills

AhmadAlamoush committed:
perf(pathPlanner): Fix undefined skill error

Check if the skill variable is "undefined"

Elscha committed:
chore: Strict checks enabled
